apps/labs/posts/array-api-support-scikit-learn.md

```diff
@@ -0,0 +1,144 @@
+hero:
+  src: /posts/array-api-support-scikit-learn/data-api-scikit-learn-logos-hero.png
+  alt: 'The Data API logo next to the scikit-learn logo.'
+---
+
+The Consortium for Python Data API Standards developed the Array API standard, which aims to define consistent behavior between the ecosystem of array libraries, such as PyTorch, NumPy, and CuPy. The Array API standard enables libraries, such as scikit-learn, to write code once with the standard and have it work on multiple array libraries. With PyTorch tensors or CuPy tensors, it is now possible to run computation on accelerators, such as GPUs.
+
+With the release of scikit-learn 1.3, we enabled experimental Array API for a limited set of machine learning models. Array API support is gradually expanding to include more machine learning models and functionality on the development branch. Scikit-learn depends on the `array_api_compat` library for Array API support. The `array_api_compat` extends the Array API standard to NumPy's, CuPy arrays, and PyTorch Tensors. In this blog post, we cover scikit-learn's public interface for enabling Array API, the performance gain of running on an accelerator, and the challenges we faced when integrating Array API.
+
+## Benchmarks
+Scikit-learn was initially developed to run on CPUs with NumPy arrays. With Array API support, a limited set of scikit-learn models and tools can now run with other array libraries and devices like GPUs. The following benchmark results are from running scikit-learn's LinearDiscurementAnalysis with different array libraries:
+
+![Bar plot with benchmark results comparing NumPy, PyTorch CPUs, and PyTorch GPUs running Linear Discriminant Analysis. The PyTorch GPU results are marked as 27 times faster compared to NumPy for fitting the model and 28 times faster compared to NumPy for prediction.](/posts/array-api-support-scikit-learn/benchmark_results.png)
+
+The training and prediction times are improved when using PyTorch compared to NumPy. Running the computation on PyTorch CPU tensors is faster than NumPy because PyTorch CPU operations are multi-threaded by default.
+
+## scikit-learn's Array API interface
+
+Scikit-learn extended its experimental Array API support in version 1.3 to support NumPy's `ndarrays`, CuPy's `ndararys`, and PyTorch `Tensors`. By themselves, these array objects do not implement the Array API specification. To overcome this limitation, Quansight engineer Aaron Meurer, led the development of `array_api_compat` to adopt the Array API to NumPy, CuPy, and PyTorch. Scikit-learn directly uses `array_api_compat` for its Array API support. There are two APIs enabling Array API in scikit-learn: a global configuration and a context manager. The following example use a context manager:
+
+```python
+import sklearn
+import torch
+from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
+
+X_torch_cpu, y_torch_cpu = torch.asarray(X_np), torch.asarray(y_np)
```

Comment:
`X_np` and `y_np` aren't defined, so this code doesn't run. Importing some example dataset would be good.